stumpy/scrump.py

```diff
@@ -103,67 +109,113 @@ def _compute_PI(
     for i in indices[start:stop]:
         Q = T_A[i : i + m]
         QT[:] = core._sliding_dot_product(Q, T_B)
-        # Update P[i] relative to all T[j : j + m]
         squared_distance_profile[:] = core._mass(Q, T_B, QT, μ_Q[i], σ_Q[i], M_T, Σ_T)
         squared_distance_profile[:] = np.square(squared_distance_profile)
         if excl_zone is not None:
-            zone_start = max(0, i - excl_zone)
-            zone_stop = min(l, i + excl_zone)
-            squared_distance_profile[zone_start : zone_stop + 1] = np.inf
-
-            # only for self-join
-            mask = squared_distance_profile < P_squared[thread_idx]
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
-
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
-        else:
-            j = I[thread_idx, i]
-            # Given the squared distance, work backwards and compute QT
-            QT_j = (m - P_squared[thread_idx, i] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
-                m * M_T[j] * μ_Q[i]
+            core._apply_exclusion_zone(squared_distance_profile, i, excl_zone, np.inf)
+
+        nn_idx = np.argmin(squared_distance_profile)
+        core._shift_insert_at_index(
```

Comment:
In self-join, in each iteration, we use the distance profile to update top-k values for all subsequences (see the last if-block). So, when we get to this line in the `i`-th iteration, all top-k values (of the subset at index `i`) might be already filled in and we should update it (which means we should shift the values). 

It is worthwhile to note that the function's goal is to give us an approx matrix profile. So, I THINK we are not obligated to keep the best top-k! For instance, with `k=5`, the top-k values of the `i`-th sequence might be already `[0.1, 0.2, 0.3, 0.4, 0.5]` before this line. So, let's say `squared_distance_profile[nn_idx]` is 0.01, then I THINK it is NOT wrong to set the first element to 0.01, hence: `[0.01, 0.2, 0.3, 0.4, 0.5]`. It is still approximate and the top-1 approx SHOULD BE the same as the top-1 approx of the original version.

Reply:
I see! Thanks for the clarification. I wonder if it might be useful to add a comment along the lines of:
```
# It is more than likely that the top-k values for the `i`-th subsequence will already be populated (by subsequences to the left of `i`) so we must shift-insert here 
```

Reply:
Sure. That would be a good idea. 

I was wondering if it would be okay to remove "(by subsequences to the left of `i`)". I am not 100% sure but I think we are not traversing the main sequence from left to right in `prescrump`.

```
# in function prescrump

indices = np.random.permutation(range(0, l, s)).astype(np.int64) # line 417
P, I = _prescrump(...) # line 418
``` 

So, I think we might check out a subsequence with start index that is greater than `i` before the `i` itself (and its distance profile can help us update top-k of the `i`-th subsequence!)  

What do you think?

Reply:
This is what I think:
```
# input: 
# T with length n=12
# m = 3 
# l = n - m + 1

>>> range(0, l, s)
0, 3, 6, 9  # shown in color red in image below

indices = np.random.permutation(range(0, l, s)).astype(np.int64) # line 417
# for instance, 
# indices: 6, 0, 3, 9  
```

So, we start with subseq at index 6, and its nearest neighbor is at index 2. We then use `QT-related for-loop`  and we move forward (blue arrows) and move backward (green arrows). In blue direction, we calculate `dist(seq7, seq3)` and `dist(seq8, seq4)`. With the former, i.e. `dist(seq7, seq3)`, we can update top-k values in `P[7, :]`, and in `P[3, :]`. 


![image](https://user-images.githubusercontent.com/38519522/179281988-a2d4b482-0186-4a48-8692-242bd2a70417.png)

So, I think it is possible that we populate the top-k values for sample index `3` with help of other samples that might be even on its right, e.g. index `6`. Furthermore, the distance profile of index `6` can also update those top-k values (and top-k values of all other non-trivial neighbors)  as well.

Please let me  know if I missed something...

Reply:
**off-topic note: Speeding up `scrump` calculation**

I was thinking more about this and I think it might be possible to speed up the `scrump` calculation when parameter `pre_scrump` is set to True

If `scrump(..., pre_scrump=True)`, that means we use `prescrump` first to get an approximate matrix profile. Now, we can have a 2D Boolean array `visited` that can be returned at the end of `prescrump` process. 

`visited[i, j]` is True if the distance between `subseq i` and `subseq j` were calculated throughout the prescrump process. For instance, for sample index `6`, we can set `visited[6, :] = True` because we used its distance profile already to update the matrix profile! (In other words, we calculate the distances between `6` and all of its neighbors!) 

Also, `visited[7, 3]` and `visited[8, 4]` should be true as well! 

So, in `.update()`, we only calculate distance between x and y if `visited[x, y]=False`.

What do you think?! I can try it to see its impact on the computing time.

Reply:
Hmm, I see! Basically, we need to check that the index does not exist in `I` yet. Interesting

Reply:
> Would you mind creating an issue for this so that we can document and discuss?

I opened issue #643 

> Can we come up with an input that would cause a failing unit test that covers this situation?

I will think about it

Reply:
@seanlaw 
> Can we come up with an input that would cause a failing unit test that covers this situation?

I caught it in our existing input. I added the test:
```
# in tests/naive.py
def check_no_duplicate(I):
    """
    check duplicates (excluding `-1`) in the matrix profile indices. 
    It returns True if no duplicate is found
    """
    flag = True
    for i in range(I.shape[0]):
        mask = I[i] >= 0 # ignore the -1 values
        flag = flag and (len(set(I[i, mask])) == np.sum(mask))

    return flag


# in tests/test_scrump.py
# ADD assert naive.check_no_duplicate(comp_I) for KNN test functions
```
---

**And here is the error...**
```
================================== FAILURES ===================================
___________________ test_prescrump_self_join_KNN[T_A0-T_B0] ___________________

T_A = array([ 9.0e+00,  8.1e+03, -6.0e+01,  7.0e+00])
T_B = array([ 584.,  -11.,   23.,   79., 1001.,    0.,  -19.])

    @pytest.mark.parametrize("T_A, T_B", test_data)
    def test_prescrump_self_join_KNN(T_A, T_B):
        m = 3
        zone = int(np.ceil(m / 4))
        for k in range(2, 4):
            for s in range(1, zone + 1):
                seed = np.random.randint(100000)

                np.random.seed(seed)
                ref_P, ref_I = naive.prescrump(T_B, m, T_B, s=s, exclusion_zone=zone, k=k)

                np.random.seed(seed)
                comp_P, comp_I = prescrump(T_B, m, s=s, k=k)

                npt.assert_almost_equal(ref_P, comp_P)
                npt.assert_almost_equal(ref_I, comp_I)
>               assert naive.check_no_duplicate(comp_I)
E               assert False
E                +  where False = <function check_no_duplicate at 0x00000275790F7D00>(array([[4, 4],\n       [3, 3],\n       [0, 0],\n       [1, 1],\n       [0, 0]], dtype=int64))
E                +    where <function check_no_duplicate at 0x00000275790F7D00> = naive.check_no_duplicate

tests\test_scrump.py:682: AssertionError
=========================== short test summary info ===========================
FAILED tests/test_scrump.py::test_prescrump_self_join_KNN[T_A0-T_B0] - assert...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
======================== 1 failed, 54 passed in 50.26s ========================
Error: pytest encountered exit code 1
```

As we can see, in this particular case, ALL rows of the matrix profile indices `comp_I` have duplicates!

Reply:
I don't think this should be the form of the test. For a unit test, given the inputs, do the outputs match the expected outputs? That should be the unit test. It should compare computed outputs to expected (correct) outputs given the same input.

Reply:
Right! Now that I caught the issue, I will remove the newly added function and the assert statement. I will go and change the naive version of prescrump/ scrump.

Reply:
I revised the naive version of `prescrump` to avoid duplicates in the matrix profile indices:

```
# test_prescrump_self_join_KNN[T_A0-T_B0]

T_A = array([ 9.0e+00,  8.1e+03, -6.0e+01,  7.0e+00])
T_B = array([ 584.,  -11.,   23.,   79., 1001.,    0.,  -19.])

    @pytest.mark.parametrize("T_A, T_B", test_data)
    def test_prescrump_self_join_KNN(T_A, T_B):
        m = 3
        zone = int(np.ceil(m / 4))
        for k in range(2, 4):
            for s in range(1, zone + 1):
                seed = np.random.randint(100000)

                np.random.seed(seed)
                ref_P, ref_I = naive.prescrump(T_B, m, T_B, s=s, exclusion_zone=zone, k=k)

                np.random.seed(seed)
                comp_P, comp_I = prescrump(T_B, m, s=s, k=k)

>               npt.assert_almost_equal(ref_P, comp_P)
E               AssertionError:
E               Arrays are not almost equal to 7 decimals
E
E               Mismatched elements: 5 / 10 (50%)
E               Max absolute difference: 2.9402256
E               Max relative difference: 25.27300764
E                x: array([[0.1163386, 2.982409 ],
E                      [2.6940739, 3.284127 ],
E                      [3.0000926, 3.0565642],...
E                y: array([[0.1163386, 0.1163386],
E                      [2.6940739, 2.6940739],
E                      [3.0000926, 3.0000926],...

```

So, I used a boolean 2D array to check if `dist[i, j]` is already used (to update `P[i]`) or not.
```
# in tests/naive.py

def prescrump(T_A, m, T_B, s, exclusion_zone=None, k=1):
    dist_matrix = distance_matrix(T_A, T_B, m)
    is_visited = np.full_like(dist_matrix, 0, dtype=bool) # NEW
    ###
```

However, I think this is not a good approach as we cannot afford to store this Boolean 2D array for large time series. 
Hence:
* Alternative I: if our priority is to make sure that the performance for case `k=1` is as good as the existing `main` version, we can avoid the Boolean 2D array and simply check if an index `j` is in `I[i]` or not! This is O(k), and it can become a problem when `k` becomes larger.

* Alternative II: use a way to store sparse(?) boolean 2D array `is_visited`. However, the sparsity will decrease as we execute the `.update()` method more and more. 

---

As you suggested, I will slow down. I will resume this after taking care of other scrump-related issues.

Reply:
I think we should go with "Alternative I" and create a `core._isin(a, v)` function that returns `True`/`False`. Something simple like:

```
@njit(parallel=True)
def _isin(a, v):
for i in prange(len(a)):
    if v == a[i]:
        return True

return False
```

I believe that this will be VERY fast

Reply:
Great! You are right...so, if I understand correctly, when we do `idx = np.searchsorted(..., side='rigth')`, we just need to search the slice `arr[:idx]`. Cool :)

I also need to take care of this duplicate thing in the `.update()` call (in `stumpy/scrump.py`). But, I will first take care of prescrump and then, I will work on the scrump.

Reply:
> we just need to search the slice arr[:idx]

Exactly! It's actually a very cheap calculation as there are a lot of checks happen before we insert anything. So, we'd:

1. Find the `idx` with `np.searchsorted`
2. Check if the subsequence index is in `arr[:idx]` 
3. If yes, then don't insert
4. If no, then insert

Reply:
Yes, I don't think we need a separate function. I think we can just do:

```
if not idx in I[i]: 
    # insert idx into top-k
```

Reply:
Yes! For the sake of comparison, I just converted the `for-loop` to `if v in a:`. But, I am going to use what you wrote above in the code :) Thanks :)